### ISSUE TITLE
Consultation on the game making.

### ISSUE THREAD

**Issue by 0xdream** [stievie/ABx#85]
Hello stievie:

I am puzzled on game making and would like to ask you for advice, if you have time to reply to me, I would appreciate it very much.

I am now a first year graduate student, familiar with CPP programming language, and have learned some network principles and graphics principles (write minecraft-like in OpenGL). But I don't have any real project experience. I don't know how to build a real MMORPG, because I want to write an open world game, but I always feel that I can't start. At the beginning, the function can be simple, as long as there is a basic framework. Later, I can slowly add code, you know, just like the project ABx you developed.

Now what I want to know is whether I can make changes based on your project, delete some details and leave a basic framework, so as to understand how the network communication of the whole project is carried out and some practical skills of Urho3d, and then let it become another completely new game. 

Do you think this is feasible? I really need some advice.

chunfeng_tsin,
Best regards.

**Comment by stievie** [comment]
First, I think MMORGP is a scary thing, because this means 100,000 or more concurrent players. But its unrealistic to think we can make a game which attracts so many players, so maybe we are talking about an multiplayer online RGP without the "massive".

Sure, I think it's possible, but its huge and you need a lot of time and motivation, it takes years to develop such a game. I, for example, started with this over 3 years ago, and it is nowhere close to something playable and enjoyable.

An open world game, is even more difficult to make, because it introduces some more problems that need to be solved, e.g. huge maps with many actors on the map.

I do think you could use ABx to start with, but an open world game is different to an instanced game. You maybe need to change how path finding works, how maps, terrains are handled, I don't know how all this works with really large maps. Also you may need to change some networking, introduce better interest management. It doesn't  make sense to send the whole game state with thousands of actors to every player etc.

I also think the server is tightly coupled to the type of game you want to make. Maybe that's the reason there are not many general purpose game server frameworks (one being https://github.com/kbengine/kbengine).

Long story short, if you want to make such a game, sure start with it, but don't expect to finish it in a couple of months. Use ABx to start with, if its not too different to the game you imagine.

I also never made a game, or even a game server before. I am by no means an expert in making game servers. But I started once and learn the things when I need them. So don't be overwhelmed when you don't know how to do things before, but investigate and solve the problems when they arise, one by one.

I'm not sure if that helps you, but that's what I think when I read your question.

Good luck and best wishes.

**Comment by stievie** [comment]
Oh, btw, if you find asio (the library I use for networking) confusing, and it can be because doing things asynchronously may need some other thinking, here are good introductions:
https://www.youtube.com/watch?v=2hNdkYInj4g
https://www.youtube.com/watch?v=UbjxGvrDrbw

**Comment by 0xdream** [comment]
Thanks stievie, 
These suggestions are really practical and useful.
I may be too eager to succeed. I will try them patiently and carefully. 
Thank you again!
